mypy/typeops.py

```diff
@@ -397,6 +401,11 @@ def make_simplified_union(items: Sequence[Type],
                     and is_proper_subtype(tj, item, keep_erased_types=keep_erased)
                     and is_redundant_literal_instance(item, tj)  # XXX?
             ):
+                if keep_none and not state.strict_optional and isinstance(tj, NoneType):
+                    # Sometimes we need to keep `None`, because we might
+                    # be doing branch analysis.
+                    # See: https://github.com/python/mypy/issues/11705
+                    continue
```

Comment:
Do we need the `keep_none` flag if we're checking for strict_optional and NoneType? The logic seems similar to `mypy/checkexpr.py@L2826`.

Reply:
Yes, we do. Because we only need this logic to be executed at some use-cases. If I remove this param, we will have 5-6 failing tests.